pulsar-io/hbase/src/main/java/org/apache/pulsar/io/hbase/sink/HbaseAbstractSink.java

```diff
@@ -0,0 +1,187 @@
+import java.util.concurrent.TimeUnit;
+
+/**
+ * A Simple abstract class for Hbase sink
+ * Users need to implement extractKeyValue function to use this sink
+ */
+@Slf4j
+public abstract class HbaseAbstractSink<T> implements Sink<T> {
+
+    @Data(staticConstructor = "of")
+    @Setter
+    @Getter
+    @EqualsAndHashCode
+    @ToString
+    public static class TableDefinition {
+        private final String rowKeyName;
+        private final String familyName;
+        private final List<String> qualifierNames;
+    }
+
+    private HbaseSinkConfig hbaseSinkConfig;
+    private Configuration configuration;
+    private Connection connection;
+    private Admin admin;
+    private TableName tableName;
+    private Table table;
+
+    protected TableDefinition tableDefinition;
+
+    // for flush
+    private int batchTimeMs;
+    private int batchSize;
+    private long currentTime;
+    private List<Record<T>> incomingList;
+    private ScheduledExecutorService flushExecutor;
+
+    @Override
+    public void open(Map<String, Object> config, SinkContext sinkContext) throws Exception {
+        hbaseSinkConfig = HbaseSinkConfig.load(config);
+        if (hbaseSinkConfig.getZookeeperQuorum() == null
+                || hbaseSinkConfig.getZookeeperClientPort() == null
+                || hbaseSinkConfig.getTableName() == null) {
+            throw new IllegalArgumentException("Required property not set.");
+        }
+
+        getTable(hbaseSinkConfig);
+        tableDefinition = getTableDefinition(hbaseSinkConfig);
+
+        batchTimeMs = hbaseSinkConfig.getBatchTimeMs();
+        batchSize = hbaseSinkConfig.getBatchSize();
+        currentTime = System.currentTimeMillis();
+        incomingList = Lists.newArrayList();
+        flushExecutor = Executors.newScheduledThreadPool(1);
+    }
+
+    @Override
+    public void close() throws Exception {
+        if (null != admin) {
+            admin.close();
+        }
+
+        if (null != connection) {
+            connection.close();
+        }
+
+        if (null != flushExecutor) {
+            flushExecutor.shutdown();
+        }
+    }
+
+    @Override
+    public void write(Record<T> record) throws Exception {
+        synchronized (incomingList) {
+            incomingList.add(record);
+
+            if (System.currentTimeMillis() - currentTime > batchTimeMs * 1000
+                    || incomingList.size() >= batchSize) {
+                flushExecutor.schedule(() -> flush(), 0, TimeUnit.MILLISECONDS);
+            }
+        }
+    }
+
+    private void flush() {
+        try {
+            List<Put> puts = new ArrayList<>();
+            for (Record<T> value : incomingList) {
+                bindValue(value, puts);
+            }
+
+            if (CollectionUtils.isNotEmpty(puts)) {
+                table.put(puts);
+            }
+            admin.flush(tableName);
+            incomingList.forEach(tRecord -> tRecord.ack());
+
+            incomingList.clear();
```

Comment:
you are potentially clear the records being added to incomingList since `write` is called in a different thread.
so you still need two list to swap between flushes.

Reply:
@sijie 
I modify write and flush like :
```java
@Override
    public void write(Record<T> record) throws Exception {
        synchronized (incomingList) {
            incomingList.add(record);

            if (System.currentTimeMillis() - currentTime > batchTimeMs * 1000
                    || incomingList.size() >= batchSize) {
                flushExecutor.schedule(() -> flush(), 0, TimeUnit.MILLISECONDS);
            }
        }
    }

    private void flush() {
        List<Record<T>> swapList = null;
        List<Put> puts = new ArrayList<>();
        try {
            if (incomingList.size() > 0) {
                swapList = incomingList;
                incomingList = Lists.newArrayList();

                for (Record<T> value : swapList) {
                    bindValue(value, puts);
                }
            }
            
            if (CollectionUtils.isNotEmpty(puts)) {
                table.put(puts);
            }
            admin.flush(tableName);
            if (CollectionUtils.isNotEmpty(swapList)) {
                swapList.forEach(tRecord -> tRecord.ack());
            }
            currentTime = System.currentTimeMillis();
        } catch (Exception e) {
            log.error("Got exception ", e);
            if (CollectionUtils.isNotEmpty(swapList)) {
                swapList.forEach(tRecord -> tRecord.fail());
            }
        }
    }
```

Reply:
yeah yes. I made a few inline comments, it is same as what you proposed in the comment.